text/0253-cdk-metadata-v2.md

```diff
@@ -0,0 +1,267 @@
+  which one the class (`object.constructor`) is in. If the class is non-exported
+  (such as an `class Import implements IBucket { ... }`) it will not show up though, and we
+  won't be able to figure out what file it's from.
+  * A constructor may be in multiple module's `exports` (because of
+    re-exporting). If it is, take the one with the most qualified filename.
+* We crawl up the inheritance hierarchy to find the first class that is exported in this way.
+  For unrecognized classes, we'll end up seeing `Construct`.
+* We can find the package a class is from by crawling up the directory tree from the filename
+  we identified and looking for `package.json`, getting the `name` and `version` from that file.
+
+### Submodule names
+
+The only thing we can't get from this is a qualified "submodule" name which is going to
+become relevant in the monolibrary v2, where both the EKS and ECS submodules have a `Cluster` class,
+for example, which are otherwise indistinguishable apart from the filename. We also cannot
+completely rely on the v2 (monocdk) directory structure, as this lookup also needs to work
+for non-v2 AWS-vended libraries such as the `@aws-solutions-constructs` libraries.
+
+Class name | File name | Desired identifier
+--|--|--
+Cluster | .../node_modules/aws-cdk-lib/lib/aws-ecs/lib/cluster.js | aws-cdk-lib.aws_ecs.Cluster@2.1.0
+Cluster | .../node_modules/aws-cdk-lib/lib/aws-eks/lib/cluster.js | aws-cdk-lib.aws_eks.Cluster@2.1.0
+LambdaToSqs | .../node_modules/@aws-solutions-constructs/aws-lambda-sqs/lib/index.js | @aws-solutions-constructs/aws-lambda-sqs.LambdaToSqs@1.63.0
+
+It would be nice if we used submodule naming that's consistent with the jsii
+naming. There is nothing in the jsii manifest that could reliably be used for
+this, but we can use an additional lookup table that can be generated by the
+monocdk/v2 build tool, as that is in charge of the submodule names anyway.
+
+There will be a `submodules.json` next to `package.json` with the following structure:
+
+```
+{
+  "lib/aws-ecs": "aws_ecs",
+  "lib/aws-eks": "aws_eks",
+  ...
+}
+```
+
+Which is used to map subdirectories to submodule names. Absence of this file indicates no submodule names
+need to be inserted.
+
+The first matching directory found will be used, so should we ever need a
+situation to support submodules-in-submodules, we can make use of the fact
+that JSON dictionaries are ordered and make sure the deeper directories occur
+first.
+
+> Implementation note: looking up the source file from a class is a potentially
+> expensive operation as it takes a linear scan through all sources for every class.
+> We should build a reverse index in order to speed this up.
+
+## Metadata resource encoding
+
+A rough estimate of the maximum number of constructs used in a stack will be
+about 400 (maximum of 200 L1s with a corresponding L2 wrapper). We will
+therefore need to encode roughly 400 strings of the form
```

Comment:
That's the worst case, right? We aren't encoding all 400 in each stack?

Reply:
That's the worst case, correct.